server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -247,33 +255,78 @@ private static String getExceptionMessage(Throwable e) {
+                    // and wait for them. The reason is concurrent requests. Request r1 which has new field f triggers a
+                    // mapping update. Assume that that update is first applied on the primary, and only later on the replica
+                    // (it’s happening concurrently). Request r2, which now arrives on the primary and which also has the new
+                    // field f might see the updated mapping (on the primary), and will therefore proceed to be replicated
+                    // to the replica. When it arrives on the replica, there’s no guarantee that the replica has already
+                    // applied the new mapping, so there is no other option than to wait.
+                    throw new TransportReplicationAction.RetryOnReplicaException(indexShard.shardId(),
+                        "Mappings are not available on the replica yet, triggered update: " + result.getRequiredMappingUpdate());
                 }
+                location = result.getTranslogLocation();
+            }
+            return new WriteReplicaResult<>(request, location, null, indexShard, logger);
+        }
+
+        RelationName relationName = RelationName.fromIndexName(indexName);
+        DocTableInfo tableInfo = schemas.getTableInfo(relationName, Operation.INSERT);
+        var mapperService = indexShard.mapperService();
+        Function<ColumnIdent, FieldType> getFieldType = column -> mapperService.getLuceneFieldType(column.fqn());
+        TransactionContext txnCtx = TransactionContext.of(request.sessionSettings());
+
+        // Refresh insertColumns References from cluster state because ObjectType
+        // may have new children due to dynamic cluster state updates
+        // Not doing this would result in indefinite `Mappings are not available on the replica yet` errors below
+        List<Reference> targetColumns = Stream.of(insertColumns)
+            .map(ref -> {
+                Reference updatedRef = tableInfo.getReference(ref.column());
+                return updatedRef == null ? ref : updatedRef;
+            })
+            .toList();
+        Indexer indexer = new Indexer(
+            indexName,
+            tableInfo,
+            txnCtx,
+            nodeCtx,
+            getFieldType,
+            targetColumns,
+            null
+        );
+        for (ShardUpsertRequest.Item item : request.items()) {
+            if (item.seqNo() == SequenceNumbers.SKIP_ON_REPLICA) {
                 continue;
             }
-            SourceToParse sourceToParse = new SourceToParse(
-                indexShard.shardId().getIndexName(),
-                item.id(),
-                item.source(),
-                XContentType.JSON
-            );
+            long startTime = System.nanoTime();
+            ParsedDocument parsedDoc = indexer.index(item);
+            if (!parsedDoc.newColumns().isEmpty()) {
+                // this forces clearing the cache
+                schemas.tableExists(relationName);
```

Comment:
Isn't this a bit hacky? couldn't we have some dedicated method to invalidate the cache instead?

Reply:
Also not insisting, we can keep as is.